### DISCUSSION TITLE
In Internationalization routes, how to navigate to a different locale but on the same page?

### DISCUSSION THREAD

**Discussion by nonoumasy** [vercel/next.js#36965] [Help] [answered]
localeDetection: false,Anyone implemented Internationalization with Nextjs?
I'm working on implementing this right now for my site:
https://history-maps.com/

I'm using Next js Internationalized Routing:
https://nextjs.org/docs/advanced-features/i18n-routing

Spanish:
https://history-maps.com/es/story/Grand-Principality-of-Serbia

Arabic:
https://history-maps.com/ar/story/Grand-Principality-of-Serbia

I have a language selector on the header, but if I'm not at the home url, every time I switch to another languages, it navigates back to the home url. Is there a way with useRouter I can stay on the same page but just switch the locale ?
```js
export const LanguageSelector = () => {
  const router = useRouter()
  const handleChange = (e) => {
    router.push("/", "/", { locale: e.target.value })
  }

  return (
    <select className="select-language" onChange={handleChange}>
      <option value="en">english</option>
      <option value="es">español</option>
      <option value="pt">Português</option>
      <option value="ru">русский</option>
      <option value="ar">العربية</option>
      <option value="zh">中文</option>
      <option value="ja">日本語</option>
      <option value="hi">नहीं</option>
    </select>
  )
}
```

**Comment by GabenGar** [comment]
Instead of the form input use the list of `<Link>`s.
You can look at [site](https://gabengar.vercel.app) and the [source code](https://github.com/GabenGar/next-js-site/blob/master/src/components/language/switcher.tsx) of the component.
Also think twice about showing language selector in a single language, since it makes harder to switch from it for non-native speakers.

**Comment by arielvieira** [comment, marked as answer]
Yes, there is. You can use `router.pathname` and `router.asPath`(for dynamic paths) when calling `router.push` to handle that case.

Example:
```js
  const handleChange = (e) => {
    router.push(router.pathname, router.asPath, { locale: e.target.value })
  }
```

**Reply by icyJoseph** [reply]
This is the absolute way to do it. 

You could also take the `locales` list from `useRouter` and map over that (after doing an undefined check), just to make sure that you show locales which are actually supported. Though this involves creating a table to look up their description, and ordering.
